acceptance/tests/resource/exec/should_run_command_in_cwd.rb

```diff
@@ -2,117 +2,180 @@
     end
 
-    step "Defaults to the PWD if the CWD is not provided" do
-      pwd = on(agent, print_cwd).stdout.chomp
-      apply_manifest_on(agent, exec_resource_manifest.call) do |result|
-        assert_match(to_regex(pwd), result.stdout, 'The Exec resource does not default to using the PWD when the CWD is not provided')
+    step "Runs the command in the user specified CWD" do
+      apply_manifest_on(agent, exec_resource_manifest("#{echo_to} cwd_test2", {:cwd => testdir, :path => path}), :catch_failures => true)
+      test_file(agent, File.join(testdir, 'cwd_test2'), 'Exec did not create file in test directory, exec resource not using :cwd given')
+    end
+
+    step "Errors if the user specified CWD does not exist" do
+      apply_manifest_on(agent, exec_resource_manifest("#{echo_to} cwd_test3", {cwd: non_existant_dir, :path => path}), :expect_failures => true) do |result|
+        assert_equal(4, result.exit_code, "Exec manifest still executed with non-existant :cwd")
       end
     end
 
-    tmpdir = mk_tmpdir.call      
+    # "onlyif" testing will require some form of runnable test in the testdir for the
+    # onlyif clause to actually execute. The runnable test we will use is attempting to
+    # 'cat' an unqualified file that will only exist in the testdir
+    create_remote_file(agent, File.join(testdir, 'testdir_onlyif.txt'), 'testing')
+
+    step 'Runs a "check" command (:onlyif or :unless) in the user specified CWD' do
+      apply_manifest_on(agent, exec_resource_manifest("#{echo_to} cwd_test4", {cwd: testdir, :path => path, :onlyif => "#{cat} testdir_onlyif.txt"}), :expect_changes => true)
+      test_file(agent, File.join(testdir, 'cwd_test4'), 'Exec did not create file in test directory, exec resource not using :cwd given')
+    end
 
-    step "Runs the command in the specified CWD" do
-      apply_manifest_on(agent, exec_resource_manifest.call(cwd: tmpdir)) do |result|
-        assert_match(to_regex(tmpdir), result.stdout, 'The Exec resource does not let you specify the CWD')
+    step 'Does not run the exec if the "check" command (:onlyif or :unless) fails' do
+      apply_manifest_on(agent, exec_resource_manifest("#{echo_to} cwd_test5", {cwd: testdir, :path => path, :onlyif => "foobar"}), :expect_failures => true) do |result|
+        assert_equal(4, result.exit_code, "Exec manifest still executed with failed :onlyif clause")
       end
     end
 
-    step "Errors if the CWD does not exist" do
-      cwd = agent.platform =~ /windows/ ? 'C:\nonexistent_dir' : '/nonexistent_dir'
-      apply_manifest_on(agent, exec_resource_manifest.call(cwd: cwd)) do |result|
-        assert_match(to_regex(cwd), result.stderr, 'The Exec resource does not error when the CWD does not exist!')
+    tmpdir_noaccess = agent.tmpdir("mock_dir")
+    create_remote_file(agent, File.join(tmpdir_noaccess, 'noaccess.txt'), 'foobar')
+    username = "pl#{rand(999999).to_i}"
+
+    step "Setup user for 'no access' test" do
```

Comment:
I think it would be better to leave a long comment describing these tests for each platform instead of commenting within the code snippets. So I recommend something like
```
The next three steps are so we test X. On Solaris, we test this by doing ... On Windows ... On OSX ...
```

and then don't comment the code snippets. This way, people have a better idea of what's going on from the get-go rather than having to piece it together from the comments.